pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -57,10 +65,213 @@ func newCRIStatsProvider(
+		if imageFsInfo != nil {
+			// The image filesystem UUID is unknown to the local node or
+			// there's an error on retrieving the stats. In these cases, we
+			// omit those stats and return the best-effort partial result. See
+			// https://github.com/kubernetes/heapster/issues/1793.
+			s.AvailableBytes = &imageFsInfo.Available
+			s.CapacityBytes = &imageFsInfo.Capacity
+			s.InodesFree = imageFsInfo.InodesFree
+			s.Inodes = imageFsInfo.Inodes
+		}
+		return s, nil
+	}
+
+	return nil, fmt.Errorf("imageFs information is unavailable")
+}
+
+// getFsInfo returns the information of the filesystem with the specified
+// storageID. If any error occurs, this function logs the error and returns
+// nil.
+func (p *criStatsProvider) getFsInfo(storageID *runtimeapi.StorageIdentifier) *cadvisorapiv2.FsInfo {
+	if storageID == nil {
+		glog.V(2).Infof("Failed to get filesystem info: storageID is nil")
+		return nil
+	}
+	fsInfo, err := p.cadvisor.GetFsInfoByFsUUID(storageID.Uuid)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to get the info of the filesystem with id %q: %v", storageID.Uuid, err)
+		if err == cadvisorfs.ErrNoSuchDevice {
+			glog.V(2).Info(msg)
+		} else {
+			glog.Error(msg)
+		}
+		return nil
+	}
+	return &fsInfo
+}
+
+func (p *criStatsProvider) makePodStats(podSandbox *runtimeapi.PodSandbox) *statsapi.PodStats {
+	s := &statsapi.PodStats{
+		PodRef: statsapi.PodReference{
+			Name:      podSandbox.Metadata.Name,
+			UID:       podSandbox.Metadata.Uid,
+			Namespace: podSandbox.Metadata.Namespace,
+		},
+		// The StartTime in the summary API is the pod creation time.
+		StartTime: metav1.NewTime(time.Unix(0, podSandbox.CreatedAt)),
+		// Network stats are not supported by CRI.
+	}
+	podUID := types.UID(s.PodRef.UID)
+	if vstats, found := p.resourceAnalyzer.GetPodVolumeStats(podUID); found {
+		s.VolumeStats = vstats.Volumes
+	}
+	return s
+}
+
+func (p *criStatsProvider) makeContainerStats(
+	stats *runtimeapi.ContainerStats,
+	container *runtimeapi.Container,
+	rootFsInfo *cadvisorapiv2.FsInfo,
+	uuidToFsInfo map[runtimeapi.StorageIdentifier]*cadvisorapiv2.FsInfo,
+) *statsapi.ContainerStats {
+	result := &statsapi.ContainerStats{
+		Name: stats.Attributes.Metadata.Name,
+		// The StartTime in the summary API is the container creation time.
+		StartTime: metav1.NewTime(time.Unix(0, container.CreatedAt)),
+		CPU:       &statsapi.CPUStats{},
+		Memory:    &statsapi.MemoryStats{},
+		Rootfs:    &statsapi.FsStats{},
+		Logs: &statsapi.FsStats{
+			Time:           metav1.NewTime(rootFsInfo.Timestamp),
+			AvailableBytes: &rootFsInfo.Available,
+			CapacityBytes:  &rootFsInfo.Capacity,
+			InodesFree:     rootFsInfo.InodesFree,
+			Inodes:         rootFsInfo.Inodes,
+			// UsedBytes and InodesUsed are unavailable from CRI stats.
+			//
+			// TODO(yguo0905): Get this information from kubelet and
+			// populate the two fields here.
+		},
+		// UserDefinedMetrics is not supported by CRI.
+	}
+	if stats.Cpu != nil {
```

Comment:
You can use stas.GetCpu().GetUsageCoreNanoSeconds().GetValue() to get rid of all the nil checks.

Reply:
To avoid setting default values in the results, I still need to check whether `stas.GetCpu().GetUsageCoreNanoSeconds().GetValue()` is 0 or not.